### DISCUSSION TITLE
Make basic_foo_encoder instances reusable

### DISCUSSION THREAD

**Discussion by ecorm** [danielaparker/jsoncons#378] [Ideas]
There is currently no way to reuse `basic_foo_encoder` instances after they've encoded a document. It would be nice to reuse the same encoder in a networking application to avoid a bunch of allocations and cache misses when there are many outbound messages to encode in a long-lived program.

I suggest a `reset` method that takes a new `Sink` and resets all internal state so that it may encode a new document.

**Comment by danielaparker** [comment]
That's certainly an option. 

basic_xxx_cursor could have a reset function as well. Perhaps with or without a new source, without if there are more documents in the original source.

**Reply by ecorm** [reply]
The proposed `basic_foo_encoder::reset` could also have overloads that take or don't take a new `Sink`. The one without the new `Sink` would be for when multiple documents are to be output to the same sink.

I might be willing to submit a PR for this if you're open to that sort of thing. If you prefer to do this yourself, I understand (that would be my preference as well, to be honest).

**Reply by danielaparker** [reply]
Feel free to submit a PR for this change. The recently added `basic_cbor_cursor2` and `basic_msgpack_cursor2`, following the earlier discussion, also need this change.

**Reply by ecorm** [reply]
I'm starting on this today.

**Reply by ecorm** [reply]
Please check and comment on my first jab at implementing this at https://github.com/ecorm/jsoncons/commit/5133473c46200e7deaf80a776337333fd6bf85f6

I will add unit tests and will also make the same enhancements to the other codecs. I'll submit a PR when I'm done.

**Reply by danielaparker** [reply]
Off hand it looks okay to me. The parameter to `initialize_with_string_view` might better match it's name if it were rewritten as
```
 void initialize_with_string_view(string_view_type sv, std::error_code& ec)
```
and source forwarded into it in the constructor, but either is fine with me.

**Reply by ecorm** [reply]
When I follow the instructions in README for running tests under Linux, it ends up not running any tests at all. Here's the output of `ctest --verbose`:
```
UpdateCTestConfiguration  from :/home/emile/Dev/jsoncons/build/DartConfiguration.tcl
Parse Config file:/home/emile/Dev/jsoncons/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/emile/Dev/jsoncons/build/DartConfiguration.tcl
Parse Config file:/home/emile/Dev/jsoncons/build/DartConfiguration.tcl
Test project /home/emile/Dev/jsoncons/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: jsoncons

1: Test command: /home/emile/Dev/jsoncons/build/test/unit_tests "WORKING_DIRECTORY" "/home/emile/Dev/jsoncons/test"
1: Test timeout computed to be: 1500
1: Filters: WORKING_DIRECTORY /home/emile/Dev/jsoncons/test
1: No test cases matched 'WORKING_DIRECTORY/home/emile/Dev/jsoncons/test'
1: ===============================================================================
1: No tests ran
1: 
1/1 Test #1: jsoncons .........................   Passed    0.00 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.00 sec
```

I am fairly well-versed in CMake, but not CTest. I've managed to run tests one source file at a time using my IDE's CTest integration feature.

**Reply by danielaparker** [reply]
In any  case the tests will run automatically once you've submitted the pull request and before I accept it.

**Reply by ecorm** [reply]
In case it matters, my system's CMake version is 3.16.3. My IDE (QtCreator) uses its own CMake version 3.19.2.

**Reply by ecorm** [reply]
I figured out a way to run all tests from the command line without having to mess around with CTest. I simply ran `../build/test/unit_tests` with `jsoncons/test` as the working directory.

**Comment by ecorm** [comment]
@danielaparker,  I've completed the unit tests for resettable parsers/cursors/encoders for JSON/CBOR/Msgpack, and they all pass, as well as all other existing unit tests. I'd like for you to review what I have now before proceeding with BSON, CSV, and UBJSON. I don't actually have a need for BSON/CSV/UBJSON, but I suppose you'll also want them to be resettable for the sake of completeness and uniformity.

I can submit a PR for what I have now, and you can either merge it (after we make necessary changes), or we can just use it for discussion purposes until I submit a larger PR for all codecs. Please let me know how you'd like to proceed.

**Reply by ecorm** [reply]
@danielaparker , I'm going ahead and submitting a PR for what I have with the resettable parsers/cursors/encoders for JSON/CBOR/Msgpack. I leave it up to you to decide if you want to use that PR for commenting purposes only, or to ultimately merge it before we proceed with the BSON/CSV/UBJSON codecs.

I hope you can understand that if there are changes to be made, I'd prefer to know about them now before starting work on the remaining codecs. Having to make the same changes to all 6 codecs would be tedious! :smile: 

I'm guessing that you might on vacation on this holiday weekend, so I'll move on to something else in my work in the meantime.

**Comment by ecorm** [comment]
Completed in 8940235dd7ff1b08230197002ab31aee03a5ac73 for JSON/CBOR/Msgpack.
